(Feature) Unique Input Async Validator

sfount opened the pull request:
This pull request implements https://github.com/IMA-WorldHealth/bhima-2.X/issues/220.  
### Example (taken from source documentation)

``` js
@example 
<input id="email" type="email" bh-unique="/users/validateEmail">
```

`bh-unique` - the server end point that will verify if the value passed (through the input model) is unique 

The server route should implement a very simple API that returns either true or false given the input model:
`GET /example/checkInputValue/:value`
`true` - the value has been found
`false` - the value has not been found
### Code reference

``` html
                    <input 
                    id="new-hospital-number"
                    name="hospitalNumber"
                    class="form-control"
                    autocomplete="off"
                    ng-model="PatientRegCtrl.medical.hospital_no"
                    bh-unique="/patients/hospital_number"
                    required>

                  <!-- This is entirely optional - it drives the in-line feedback icons on the input -->
                  <span ng-if="details.hospitalNumber.$error.unique" class="glyphicon glyphicon-remove form-control-feedback" id="unique-error-icon"></span>
                  <span ng-if="details.hospitalNumber.$pending.unique"  class="glyphicon glyphicon-hourglass form-control-feedback"></span>

                  <!-- ng-message feedback -->
                  <div class="help-block" ng-messages="details.hospitalNumber.$error" ng-show="details.$submitted || details.hospitalNumber.$error.unique">
                    <p ng-messages-include="partials/templates/messages.tmpl.html"></p> 
                  </div>
```

Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/220

---

Hi! Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](./docs/STYLEGUIDE.md)
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub)
that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process
and help build a stronger application.  Thanks!

sfount committed:
(Feature) Added base bhUnique directive + service

Added files:
`directives/bhUnique.js`
`services/UniqueValidatorService.js`

The directive binds an inputs $asyncValidator with the services $http
request to a defined server end point. Exception handling and base cases
are required.

sfount committed:
(Feature) Handle corner cases and exceptions

This commit adds validation checking, as $q does not support cascading
.catch blocks a $q.deferred object is used to catch errors in the
directives scope.

If an error is thrown outside of the knowledge of the directive (no
internet, 500 server error) the directive will set invalid with a
provided key - this key is then translated and reported through an
additional record in ng-messages.

jniles commented on client/src/js/directives/bhUnique.js:
```diff
@@ -0,0 +1,69 @@
+ * on the angluar form record. 
+ *
+ * @example 
+ * <input id="email" type="email" bh-uniqe="/users/validateEmail">
```
Important typo: `bh-uniqe` --> `bh-unique`.

sfount committed:
(Chore) Fixed typo in documentation example

sfount committed:
(Example) Implement bh-unique w/ hospital no.

This commit provides an example for the bh-unique directive. Patient
registration hospital numbers are required to be unique by adding the
attribute:
`bh-unique="/patients/hospital_number"`

The previous (hardcoded) version of async validation on hospital numbers
has been removed.

A proposal for the `exists` API powering the unique directive/ service
has been proposed. It is as follows:
`GET /entity/attribute/:id/exists`

More information and examples can be found in the services documenation.

Updated the /patients route with the proposed API pattern.

sfount committed:
(Test)Additional cases for hospital no. exists API

Added 2 test cases to cover /patients/hospital_number/:id/exists
states - true and false.

* `true` - provide an existing hospital number and expect status 200
  with value of true
* `false` - provide a unique (not yet registered) hospital number
  and expect status 200 with value of false
